rfc/rfc-54/rfc-54.md

```diff
@@ -0,0 +1,175 @@
+* Hudi should infer partition field if users provide
+  as `spark.write.format("hudi").partitionBy(field)`.
+* Users need not pass all the configurations in each write operation. Once the
+  table has been created, most table configs do not change, e.g. table name
+  needs to be passed in every write, even though its only needed first time.
+  Hudi should fetch from table configs when options are not provided by the
+  user.
+
+#### Good defaults
+
+* Default values for configurations should be optimized for simple bulk load
+  scenario e.g. by default if we have NONE sort mode then it's as good as
+  parquet writes with some additional work for meta columns.
+* Make reasonable assumptions, such as do not rely on any external system (e.g.
+  hbase) for default. As another example, enable schema reconciliation by
+  default instead of failing writes.
+
+#### Reuse and consistency
+
+* Keep spark-sql and spark datasource and deltastreamer configs in sync as much
+  as possible. Document exceptions, e.g. key generator for sql is
+  ComplexKeyGenerator while for datasource it is SimpleKeyGenerator.
+* Rename/reuse existing datasource keys that are meant for same purpose.
+* In all these changes, we should support backward compatibility.
+
+#### Refactor Hive Sync
+
+* Reduce the number of configs needed for Hive sync, e.g. table name once
+  provided at the time of first write can be reused for hive sync table name
+  config as well.
+* Revisit the class hierarchy and refactor if needed.
+
+#### Configuration Builders
+
+* Users should be able to use the config builders instead of specifying config
+  keys,
+  e.g. `spark.write.format("hudi").options(HoodieClusteringConfig.Builder().withXYZ().build())`
+
+### Table APIs
+
+These APIs are meant for programmatically interacting with Hudi tables. Users
+should be able to create or update the tables using static methods.
+
+| Method Name   | Description   |
+| ------------- | ------------- |
+| bootstrap     | Create a Hudi table from the given parquet table  |
+| create        | Create a Hudi table with the given configs if it does not exist.   |
+| update        | Update rows in a Hudi table that match the given condition with the given update expression   |
+| drop          | Drop the given Hudi table completely  |
+| truncate      | Delete data from the given Hudi table but does not drop it |
+| restoreToTime | Restore Hudi table to the given older commit time  |
+
+Let's look at some examples:
+
+```java
+// create Hudi table
+HudiTable hudiTable=HudiTable.create(HoodieTableConfig.newBuilder()
+    .withTableName("tableName")
+    .withBasePath("basePath")
+    .withTableType(TableType.MERGE_ON_READ)
+    .build())
+
+// update Hudi table, add 1 to colA for all records of the current year (2022)
+hudiTable.update(
+    functions.col("dateCol").gt("2021-12-31"), // filter condition
+    functions.col("colA").plus(1) // update expression
```

Comment:
What's the running engine behind these APIs, do we have configurable/pluggable engine options ?